zpeclient.go

```diff
@@ -0,0 +1,227 @@
+package webhook
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	v1 "github.com/yahoo/k8s-athenz-istio-auth/pkg/apis/athenz/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	replacer     = strings.NewReplacer(".*", ".*", "*", ".*")
+	privateCache *Cache
+)
+
+// roleMappings - cr cache
+type roleMappings struct {
+	roleToPrincipals map[string][]*simplePrincipal
+	roleToAssertion  map[string][]*simpleAssertion
+}
+
+// simplePrincipal - principal data
+type simplePrincipal struct {
+	memberRegex *regexp.Regexp
+	expiration  time.Time
+}
+
+// simpleAssertion - processed policy
+type simpleAssertion struct {
+	resource *regexp.Regexp
+	action   *regexp.Regexp
+	effect   string
+}
+
+// Cache - cache for athenzdomains CR
+type Cache struct {
+	CrIndexInformer cache.SharedIndexInformer
+	DomainMap       map[string]roleMappings
+	lock            sync.RWMutex
+}
+
+// BuildCache - generate new athenzdomains cr cache
+func BuildCache(crIndexInformer cache.SharedIndexInformer) {
+	domainMap := make(map[string]roleMappings)
+	privateCache = &Cache{
+		CrIndexInformer: crIndexInformer,
+		DomainMap:       domainMap,
+	}
+	crIndexInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			item, ok := obj.(*v1.AthenzDomain)
+			if !ok {
+				fmt.Println("Unable to convert informer store item into AthenzDomain type.")
+				return
+			}
+			privateCache.addObj(item)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			newItem, ok := newObj.(*v1.AthenzDomain)
+			if !ok {
+				fmt.Println("Unable to convert informer store item into AthenzDomain type.")
+				return
+			}
+			privateCache.updateObj(newItem)
+		},
+		DeleteFunc: func(obj interface{}) {
+			item, ok := obj.(*v1.AthenzDomain)
+			if !ok {
+				fmt.Println("Unable to convert informer store item into AthenzDomain type.")
+				return
+			}
+			privateCache.deleteObj(item)
+		},
+	})
+}
+
+// parseData - helper function to parse AthenzDomain data and store it in domainMap
```

Comment:
Can we add a comment here saying the function is not thread safe and needs to be called with locking around it?